language/diem-vm/src/adapter_common.rs

```diff
@@ -0,0 +1,164 @@
+use crate::{counters::*, create_access_path, data_cache::StateViewCache};
+use anyhow::Result;
+use diem_state_view::StateView;
+use diem_types::{
+    account_address::AccountAddress,
+    account_config::{self, CurrencyInfoResource, RoleId},
+    transaction::{
+        GovernanceRole, SignatureCheckedTransaction, SignedTransaction, VMValidatorResult,
+    },
+    vm_status::{StatusCode, VMStatus},
+};
+use move_core_types::{
+    identifier::{IdentStr, Identifier},
+    move_resource::MoveStructType,
+};
+use move_vm_runtime::{data_cache::MoveStorage, session::Session};
+
+use crate::logging::AdapterLogSchema;
+
+/// This trait describes the VM adapter's interface.
+pub trait VMAdapter {
+    /// Creates a new Session backed by the given storage.
+    /// TODO: this probably doesn't belong in this trait. We should be able to remove
+    /// this after refactoring Diem transaction executor code.
+    fn new_session<'r, R: MoveStorage>(&self, remote: &'r R) -> Session<'r, '_, R>;
+
+    /// Checks the signature of the given signed transaction and returns
+    /// `Ok(SignatureCheckedTransaction)` if the signature is valid.
+    fn check_signature(txn: SignedTransaction) -> Result<SignatureCheckedTransaction>;
```

Comment:
Perhaps a topic for future refactorings, but it might make sense for `SignedTransaction` and `SignatureCheckedTransaction` to be associated types that implement minimal traits.

Reply:
Hmm are you talking about parameterizing  `SignedTransaction` too? so the signature of this method will look something like this:
```rust
fn check_signature<T: SignedTransactionTrait>(txn: &T) -> Result<SignatureChecked<T>>;
```
And this `SignedTransactionTrait` should probably have at least these two methods:
```rust
trait SignedTransactionTrait {
    fn verify_signature(&self) -> Result<SignatureChecked<T>>;
    fn into_raw(&self) -> RawTransaction; // It will be too hard to make RawTransaction a trait too right?
    // maybe some methods for retrieving signers too?
}
```

Reply:
Perhaps? But maybe that would be overkill. I'm also doubting my associated types suggestion and wondering if going full opaque with `Vec<u8>` might be better...

Reply:
@imsungchoi-fb is working on a new variant in `TransactionAuthenticator` with opaque signatures and public keys of type `Vec<u8>`. Should be a good place to start.